osu.Game/Screens/Play/KeyCounterDisplay.cs

```diff
@@ -33,48 +27,29 @@ public partial class KeyCounterDisplay : Container<KeyCounter>
         /// </summary>
         public readonly Bindable<bool> AlwaysVisible = new Bindable<bool>(true);
 
-        public KeyCounterDisplay()
-        {
-            InternalChild = KeyFlow = new FillFlowContainer<KeyCounter>
-            {
-                Direction = FillDirection.Horizontal,
-                AutoSizeAxes = Axes.Both,
-                Alpha = 0,
-            };
-        }
-
-        protected override void Update()
-        {
-            base.Update();
-
-            // Don't use autosize as it will shrink to zero when KeyFlow is hidden.
-            // In turn this can cause the display to be masked off screen and never become visible again.
-            Size = KeyFlow.Size;
-        }
-
         public override void Add(KeyCounter key)
         {
-            ArgumentNullException.ThrowIfNull(key);
+            if (!CheckType(key))
+                throw new ArgumentException($"{key.GetType()} is not a supported {nameof(KeyCounter)}.", nameof(key));
 
             base.Add(key);
             key.IsCounting = IsCounting;
-            key.FadeTime = key_fade_time;
-            key.KeyDownTextColor = KeyDownTextColor;
-            key.KeyUpTextColor = KeyUpTextColor;
         }
 
+        protected virtual bool CheckType(KeyCounter key) => true;
```

Comment:
This is very weird. What is this for? As best I can tell it's to avoid mixing up key counters of different drawable implementations?

Hesitantly I'd suggest to make the class a generic (as in `KeyCounterDisplay<TKeyCounter> where TKeyCounter : KeyCounter`) or something. I'm not sure it's ever going to be valid to mix these?

Reply:
I don't know what to do for this one, as I feel like there could be a lot of directions, but I don't know which one is the best.

From what I can learn from other generics in the codebase, I could separate the class into generic and non-generic base classes. 
However, forcing the new implementations to inherit the generic base only is pretty _impossible_ for this class, as making the constructor `internal` on the non-generic has pretty much no effect since new implementations will essentially be created in the same assembly.

I could also just make the base class generic without splitting anything, but then it would mean that every occurrence of `KeyCounterDisplay` would have to be replaced with `KeyCounterDisplay<KeyCounter>`, and I don't know if it's okay to do that (especially for variable declarations).

Reply:
Why does the check even need to exist?